engine/table/src/main/java/io/deephaven/engine/table/impl/sources/aggregate/UngroupedAggregateSlicedColumnSource.java

```diff
@@ -0,0 +1,530 @@
+                sameIndexRunLengths.setSize(currentIndexPosition + 1);
+
+                // Preload a chunk of rowsets (and start/end offsets if appropriate).
+                final ObjectChunk<RowSet, ? extends Values> indexes;
+                final LongChunk<? extends Values> startOffsets;
+                try (final RowSequence indexRowSequence =
+                        RowSequenceFactory.wrapRowKeysChunkAsRowSequence(rowsetKeyIndices)) {
+                    if (usePrev) {
+                        indexes = groupRowSetSource.getPrevChunk(rowsetGetContext, indexRowSequence).asObjectChunk();
+                        startOffsets = startSource != null
+                                ? startSource.getPrevChunk(startGetContext, indexRowSequence).asLongChunk()
+                                : null;
+                    } else {
+                        indexes = groupRowSetSource.getChunk(rowsetGetContext, indexRowSequence).asObjectChunk();
+                        startOffsets = startSource != null
+                                ? startSource.getChunk(startGetContext, indexRowSequence).asLongChunk()
+                                : null;
+                    }
+                }
+
+                int componentKeyIndicesPosition = 0;
+                for (int ii = 0; ii < indexes.size(); ++ii) {
+                    // Get the bucket rowset for this key
+                    final RowSet currRowSet = indexes.get(ii);
+                    Assert.neqNull(currRowSet, "currRowSet");
+                    final boolean usePrevIndex = usePrev && currRowSet.isTracking();
+                    final RowSet bucketRowSet = usePrevIndex ? currRowSet.trackingCast().copyPrev() : currRowSet;
+
+                    final long indexrowKey = rowsetKeyIndices.get(ii);
+                    final int rowPos = Math.toIntExact(currRowSet.find(indexrowKey));
+
+                    final int startPos = Math.toIntExact(startOffsets != null ? startOffsets.get(ii) : startOffset);
+                    final long size = bucketRowSet.size();
+                    final long offsetInGroup = componentKeyIndices.get(componentKeyIndicesPosition);
+                    final long start = ClampUtil.clampLong(0, size, rowPos + startPos) + offsetInGroup;
+
+                    try {
+                        final int lengthFromThisIndex = sameIndexRunLengths.get(ii);
+
+                        final WritableLongChunk<OrderedRowKeys> remappedComponentKeys =
+                                componentKeyIndicesSlice.resetFromTypedChunk(componentKeyIndices,
+                                        componentKeyIndicesPosition, lengthFromThisIndex);
+                        bucketRowSet.getKeysForPositions(
+                                LongStream.range(start, start + lengthFromThisIndex).iterator(),
+                                new LongChunkAppender(remappedComponentKeys));
+
+                        componentKeyIndicesPosition += lengthFromThisIndex;
```

Comment:
I think that what you are doing is effectively assuming that the componentKeyIndices are contiguous. You need to read each of them in turn, add the startPosition for the group within the bucket, and re-map *that* position.